src/rgw/rgw_sal_dbstore.h

```diff
@@ -232,20 +232,74 @@ class DBNotification : public Notification {
+    DBStore* store;
+    RGWZoneGroupPlacementTier tier;
+  public:
+    DBPlacementTier(DBStore* _store, const RGWZoneGroupPlacementTier& _tier) : store(_store), tier(_tier) {}
+    virtual ~DBPlacementTier() = default;
+
+    virtual const std::string& get_tier_type() { return tier.tier_type; }
+    virtual const std::string& get_storage_class() { return tier.storage_class; }
+    virtual bool retain_head_object() { return tier.retain_head_object; }
+    RGWZoneGroupPlacementTier& get_rt() { return tier; }
+  };
+
+  class DBZoneGroup : public ZoneGroup {
+    DBStore* store;
+    RGWZoneGroup group;
+    std::string empty;
+  public:
+    DBZoneGroup(DBStore* _store, RGWZoneGroup& _group) : store(_store), group(_group) {}
+    virtual ~DBZoneGroup() = default;
+
+    virtual const std::string& get_id() const override { return group.get_id(); };
+    virtual const std::string& get_name() const override { return group.get_name(); };
+    virtual int equals(const std::string& other_zonegroup) const override {
+      return group.equals(other_zonegroup);
+    };
+    /** Get the endpoint from zonegroup, or from master zone if not set */
+    virtual const std::string& get_endpoint() const override;
+    virtual bool placement_target_exists(std::string& target) const override;
+    virtual bool is_master_zonegroup() const override {
+      return group.is_master_zonegroup();
+    };
+    virtual const std::string& get_api_name() const override { return group.api_name; };
+    virtual int get_placement_target_names(std::set<std::string>& names) const override;
+    virtual const std::string& get_default_placement_name() const override {
+      return group.default_placement.name; };
+    virtual int get_hostnames(std::list<std::string>& names) const override {
+      names = group.hostnames;
+      return 0;
+    };
+    virtual int get_s3website_hostnames(std::list<std::string>& names) const override {
+      names = group.hostnames_s3website;
+      return 0;
+    };
+    virtual int get_zone_count() const override {
+      /* currently only 1 zone supported */
+      return 1;
+    }
+    virtual int get_placement_tier(const rgw_placement_rule& rule,
+				   std::unique_ptr<PlacementTier>* tier) {
+      return -1;
+    }
+  };
+
   class DBZone : public Zone {
     protected:
       DBStore* store;
       RGWRealm *realm{nullptr};
-      RGWZoneGroup *zonegroup{nullptr};
+      DBZoneGroup *zonegroup{nullptr};
       RGWZone *zone_public_config{nullptr}; /* external zone params, e.g., entrypoints, log flags, etc. */  
       RGWZoneParams *zone_params{nullptr}; /* internal zone params, e.g., rados pools */
       RGWPeriod *current_period{nullptr};
       rgw_zone_id cur_zone_id;
 
     public:
       DBZone(DBStore* _store) : store(_store) {
-        realm = new RGWRealm();
-        zonegroup = new RGWZoneGroup();
+	realm = new RGWRealm();
+        zonegroup = new DBZoneGroup(store, *(new RGWZoneGroup()));
```

Comment:
> ~DBZone() = default;

who frees this stuff?

Reply:
Good catch, will fix.

Reply:
it looks like the `*(new RGWZoneGroup())` part is still leaking here